### PR TITLE
Use domain-specific language for specifying tensor shapes.

### PR DESCRIPTION
**Pull request by jesnie**
Use domain-specific language for specifying tensor shapes.
This includes a significant rewrite of the documentation for the library, and some pylint updates.

### CODE REVIEW

File: gpflow/experimental/check_shapes/docstring.lark [hunk 1]
```diff
@@ -0,0 +1,33 @@
+docstring: docs info_fields
```
Comment (vdutor) [inline gpflow/experimental/check_shapes/docstring.lark:1]: Please a docstring explaining what settings we are configuring in this file.

File: gpflow/experimental/check_shapes/parser.py [hunk 2]
```diff
@@ -0,0 +1,289 @@
+# Copyright 2022 The GPflow Contributors. All Rights Reserved.
```
Comment (vdutor) [inline gpflow/experimental/check_shapes/parser.py:1]: I'm reluctant to add all this extra code to GPflow. Will make it harder to maintain for other people in the future. Why do we need this custom parser? What functionality would we lose by using the one in lark?

File: tests/gpflow/experimental/check_shapes/test_check_shapes.py [hunk 3]
```diff
@@ -228,3 +233,43 @@ def f(x: Input) -> Tuple[Output, Output]:
...
+            * **b** has shape [2, 4].
+
+            The second parameter.
+        :returns:
+            * **return** has shape [3, 4].
+
+            The result.
+        """
+        == f.__doc__
+    )
```
Comment (vdutor) [inline tests/gpflow/experimental/check_shapes/test_check_shapes.py:275]: [😍](https://emojipedia.org/smiling-face-with-heart-eyes/)

File: gpflow/experimental/check_shapes/__init__.py [hunk 4]
```diff
@@ -13,13 +13,135 @@
 # limitations under the License.
 
 # flake8: noqa
-"""
-Library for annotating and checking the shapes of tensors.
+"""============
+check_shapes
+============
+
+A library for annotating and checking the shapes of tensors.
```
Comment (vdutor) [inline gpflow/experimental/check_shapes/__init__.py:20]: You could also add a sentence explaining the sphinx integration?

File: gpflow/experimental/check_shapes/__init__.py [hunk 5]
```diff
@@ -13,13 +13,135 @@
...
+``[<dimension specifier 1>, <dimension specifer 2>, ..., <dimension specifier n>]``, where
+``<dimension specifier i>`` is
+one of:
+
+    * ``<integer>``, to require that dimension to have that exact size.
+    * ``<name>``, to bind that dimension to a variable. Dimensions bound to the same variable must
+      have the same size, though that size can be anything.
+    * ``*<name>`` or ``<name>...``, to bind *any* number of leading dimensions to a variable. Again,
+      multiple uses of the same variable name must match the same dimension sizes. Notice this only
+      is valid for leading dimensions.
```
[the remark this answers is not in the document]
Comment (jesnie) [reply gpflow/experimental/check_shapes/__init__.py:91]: > No mention of the keyword `None` or anonymous batch dimensions `...`, are we still planning to support this?

Yes. But I'll do unnamed / anonymous dimensions in another PR, to keep things simple.

**Review by vdutor — APPROVED** [review]

**Pull request merged**